### ISSUE TITLE
[BUG] The connection failed because the addresses could not be automatically updated between longhorn-Manager components

### ISSUE THREAD

**Issue by xiaoluhong** [longhorn/longhorn#1651]
**Chart_version**: v1.0.1

**Describe the bug**
The pods was rebuilt after the node restarted, and the **Longhorn-Manager** log showed the connection <IP>:8500 timeout.

**To Reproduce**
Because I forgot the screenshot, now I can only use my description to explain this problem, but I think this problem is quite serious.
1. The host has to be restarted due to configuration optimization or failure. 
2. After the host is restarted and the cluster is back to normal, the application has been unable to run properly. Viewing the workload log shows that the PV could not be mounted properly.
3. Longhorn was normal in terms of workload status. But going into the Longhorn UI to check the status of the volume is unhealthy.
4. Viewing the Longhorn workload log in the Rancher UI, the **Longhorn-Manager** workload kept prompting the connection <IP>:8500 timeout. After querying that the IP address is no longer available in the current cluster.

**Temporary solution**
Not sure if there were any other dependency issues, so manually recreated all the Pod associated with Longhorn, and then the PV could mount normally.

**Issue analysis**
From the logs, Longhorn-Manager kept prompting for connection timeouts, but Longhorn-Manager Pod was running normally and the container did not restart automatically. So can we add logic here to force the container to exit after connection failure and timeout? 

Sorry, I don't know much about how Longhorn-Manager (or any other component) gets the instance-Manager IP.  If the IP is automatically acquired when the container is started, the new IP should be available after the container is automatically restarted.

**Comment by yasker** [comment]
@xiaoluhong Is there any log you can provide? I think the instance manager IP might be the issue here. But we should already test this scenario multiple times, so it's better if we can take a look at the log to see what triggers the issue.

CC @khushboo-rancher

**Comment by khushboo-rancher** [comment]
Yes, we have tried this scenario multiple times and never faced this issue. 
**Scenarios we have tested**:
1. Reboot the node where the pod exists.
   **Result**:
   Instance-manager, longhorn-manager pods got recreated and no problem observed.
2. Node down and added a new node.
    **Result**:
   Pod moved to the new node and Instance-manager, longhorn-manager pods got created on the new node.

@xiaoluhong In your case, did longhorn-manager never recreated on the rebooted node? Do you know how much time it took for the node to restart roughly?

**Comment by xiaoluhong** [comment]
Hi @yasker , 

Sorry, it was very late that night, the user wanted to recover the storage as soon as possible, so we directly deleted and recreated all pods, and forgot to record the log information at that time.  However, we looked at the running logs of all the components, and the most fatal problem I located was that longhorn-manager prompted a connection `< IP >:8500` timeout, and other pods did not report a fatal error message. This is not the first time that similar problems have arisen, according to colleagues. So we recreated all the Pods associated with Longhorn. 

Hi @khushboo-rancher,

We have restarted all the worker nodes, so the Pods of the whole cluster should be affected. None of us have operated longhorn-related Pods since the node rebooted. From the workload level, the state is normal. We are restarting nodes one by one, not all at once. Some node restarts have been abnormal, causing the reboot to take a long time, possibly more than half an hour. Can you tell me about the logic of Longhorn-Manager updating the Instance-Manager ip? Thank you.

**Comment by yasker** [comment]
@xiaoluhong Next time you can use `Generate Support Bundle` at the bottom of the UI to generate the support bundle before reinstallation, so we can keep the necessary log and look into it.

This kind of issue is likely a race condition. It's hard to troubleshoot without reproducing it or have any related log.

**Comment by xiaoluhong** [comment]
Ok, I will collect the log when there is a problem again. Thank you

**Comment by andrewdavies-net** [comment]
Hi, @yasker 
I have also experienced a similar problem although its fairly rare occurrence (once every month or so).

This is just my homelab playground so by all means not particularly stable environment.

I have had this occur since 0.7 to current 1.0.1
On Rancher OS nodes & now ubuntu 18.04/20.04
Rancher based cluster 2.4.5 k8s 1.18.6


**Firewall log**
  | Aug 5 11:39:22 | DMZHOSTING | Default deny rule IPv4 (1000000103) | 10.42.3.32:45188 | 10.42.3.88:8500 | TCP:S
  | Aug 5 11:39:22 | DMZHOSTING | Default deny rule IPv4 (1000000103) | 10.42.3.32:46642 | 10.42.3.57:8500 | TCP:S
  | Aug 5 11:39:22 | DMZHOSTING | Default deny rule IPv4 (1000000103) | 10.42.3.32:59012 | 10.42.3.78:8500 | TCP:S
  | Aug 5 11:39:22 | DMZHOSTING | Default deny rule IPv4 (1000000103) | 10.42.3.32:44502 | 10.42.3.30:8500 | TCP:S

I get lots of traffic leaving the cluster and hitting the firewall interface from kubernetes internal IP addresses.
Source: longhorn-manager-9m922  1/1     Running     0          14d     10.42.3.32
Destination: Unsure, but i am certain that no currently running Pod/Service exists on that IP address anymore

All nodes are on the same subnet, i assume that the traffic leaving the cluster is only because that IP no longer exists therefore it gets routed to the node/default gateway(firewall).

**Longhorn support bundle**
https://drive.google.com/file/d/1tEbZLqeuFkHbaA2-rALle6qeFLyCYYr2/view?usp=sharing

I am not really sure what causes this issue initially, i expect node instability or some sort of environment failure. However once it happens longhorn is constantly sending traffic to a no longer existent IP. Although in many cases such as the above support bundle is it running without issues.

Bundle environment info:
Currently one pod is scaled to zero so it has one unattached volume 
One node seems to be full on storage, although i suspect this is longhorn leaving redundant replicas on the node/another issue.

A few more firewall logs to show the range of IPs which this is occurring within the cluster
Aug 5 13:15:48 | DMZHOSTING | 10.42.4.30:53932 | 10.42.4.48:8500 | TCP:S
  | Aug 5 13:15:48 | DMZHOSTING | 10.42.4.30:54466 | 10.42.4.48:8500 | TCP:S
  | Aug 5 13:15:48 | DMZHOSTING | 10.42.4.30:52740 | 10.42.4.21:8500 | TCP:S
  | Aug 5 13:15:48 | DMZHOSTING | 10.42.3.32:42794 | 10.42.3.76:8500 | TCP:S
  | Aug 5 13:15:48 | DMZHOSTING | 10.42.3.32:42760 | 10.42.3.76:8500 | TCP:S
  | Aug 5 13:15:48 | DMZHOSTING | 10.42.4.30:54268 | 10.42.4.48:8500 | TCP:S
  | Aug 5 13:15:48 | DMZHOSTING | 10.42.3.32:59398 | 10.42.4.48:8500 | TCP:S

More than happy to provide any further info you would like, 
I am currently unable to find a way to recreate this but i will have another go on a new cluster.

Regards,
Andrew

**Comment by yasker** [comment]
Keep this open

**Comment by ElisaMeng** [comment]
After upgrade from 1.0.0, we also see similar problem in one of the node. longhorn manager keep flooding logs like this:

```
time="2020-10-23T20:49:47Z" level=error msg="Failed to get the notifier of instance manager instance-manager-e-4d44ba64 before monitoring: failed to open process update stream: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
time="2020-10-23T20:49:47Z" level=warning msg="Error syncing Longhorn engine longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
time="2020-10-23T20:49:50Z" level=error msg="Failed to get the notifier of instance manager instance-manager-e-4d44ba64 before monitoring: failed to open process update stream: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
E1023 20:49:50.261243       1 engine_controller.go:209] fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host"
time="2020-10-23T20:49:50Z" level=warning msg="Dropping Longhorn engine longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4 out of the queue: fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
time="2020-10-23T20:49:53Z" level=warning msg="Error syncing Longhorn engine longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
time="2020-10-23T20:49:56Z" level=warning msg="Error syncing Longhorn engine longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
time="2020-10-23T20:49:59Z" level=warning msg="Error syncing Longhorn engine longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""
E1023 20:50:02.549217       1 engine_controller.go:209] fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host"
time="2020-10-23T20:50:02Z" level=warning msg="Dropping Longhorn engine longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4 out of the queue: fail to sync engine for longhorn-system/pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: failed to get process pvc-46e38adf-828c-4358-9d93-7979db4ab9a7-e-6bf5e1d4: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 10.42.1.14:8500: connect: no route to host\""

```

**Comment by yasker** [comment]
@ElisaMeng Assume your workloads on that node are already down, can you `kubectl delete` the `instance-manager-e-xxxx` on that node? Longhorn will recreate the instance manager and that should sort things out.

Alternatively, if you take a look at `kubectl -n longhorn-system get lhim -o yaml`, you should be able to see an `instance-manager-e-xxx` has the IP `10.42.1.14`. The name of that instance-manager-e should be the same one that on this node.

We will fix this issue for v1.1.0 release.

**Comment by yasker** [comment]
This issue should be a duplicate of https://github.com/longhorn/longhorn/issues/1880

**Comment by yasker** [comment]
@andrewdavies-net Thanks for reporting. We've recently found another case of the issue in #1880. I've checked your support bundle and found the symptoms are the same - Longhorn is trying to communicate with IP addresses doesn't exist, and those address were the IP of the instance managers once. I believe the reason is the same as I explained in https://github.com/longhorn/longhorn/issues/1880#issuecomment-708717733 .

It's also very rare to reproduce since it requires the Longhorn manager to miss a few events before getting into this state.

I will close this issue as a duplicated and we will fix the issue in v1.1.0 release.

**Comment by ElisaMeng** [comment]
@yasker Thank you so much! This seem happen quite frequently in  a 50 nodes cluster. When 1.1.0 is coming out? We love longhorn! 

BTW,  what happen when instance-manager crash/restart? will it cause data lost?

**Comment by yasker** [comment]
@ElisaMeng v1.1.0 is scheduled for next month.

Also, instance manager crash will not result in data lost. If it's the replica instance manager, you will only lose a replica temporarily; If it's the engine instance manager, it will be recreated and volume will be reattached to the node, however the pod needs to be recreated to get back the mount point. For v1.0.x, we suggested set `liveness probe` (see https://longhorn.io/docs/1.0.2/high-availability/recover-volume/ ) to trigger the pod restart; for v1.1.0, we will delete the workload pod automatically to trigger the recreation of the pod, so the service should recover automatically.

**Comment by ruizcrp** [comment]
Has this been fixed? I'm having the exact same issue with v1.4.2. and also port 8500 as described above.
